pkg/metabase/grant.go

```diff
@@ -0,0 +1,397 @@
+		PermissionGroupID:    ds.MetabaseGroupID,
+		AADPermissionGroupID: ds.MetabaseAADGroupID,
+		CollectionID:         ds.CollectionID,
+		SAEmail:              ds.Email,
+	})
+	if err != nil {
+		return err
+	}
+
+	m.waitForDatabase(ctx, dbID, datasource.Table)
+
+	if ds.MetabaseGroupID > 0 || ds.MetabaseAADGroupID > 0 {
+		err := m.client.RestrictAccessToDatabase(ctx, []int{ds.MetabaseGroupID, ds.MetabaseAADGroupID}, dbID)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := m.HideOtherTables(ctx, dbID, datasource.Table); err != nil {
+		return err
+	}
+
+	if err := m.client.AutoMapSemanticTypes(ctx, dbID); err != nil {
+		return err
+	}
+
+	m.log.Infof("Created Metabase database: %v", ds.Dataset.Name)
+	return nil
+}
+
+func (m *Metabase) createServiceAccount(ds *models.Dataset) ([]byte, string, error) {
+	projectResource := os.Getenv("GCP_TEAM_PROJECT_ID")
+	request := &iam.CreateServiceAccountRequest{
+		AccountId: "nada-" + MarshalUUID(ds.ID),
+		ServiceAccount: &iam.ServiceAccount{
+			Description: "Metabase service account for dataset " + ds.ID.String(),
+			DisplayName: ds.Name,
+		},
+	}
+
+	account, err := m.iamService.Projects.ServiceAccounts.Create("projects/"+projectResource, request).Do()
+	if err != nil {
+		return nil, "", err
+	}
+
+	iamPolicyCall := m.crmService.Projects.GetIamPolicy(projectResource, &cloudresourcemanager.GetIamPolicyRequest{})
+	iamPolicies, err := iamPolicyCall.Do()
+	if err != nil {
+		return nil, "", err
+	}
+
+	iamPolicies.Bindings = append(iamPolicies.Bindings, &cloudresourcemanager.Binding{
+		Members: []string{"serviceAccount:" + account.Email},
+		Role:    "projects/" + projectResource + "/roles/nada.metabase",
+	})
+
+	iamSetPolicyCall := m.crmService.Projects.SetIamPolicy(projectResource, &cloudresourcemanager.SetIamPolicyRequest{
+		Policy: iamPolicies,
+	})
+
+	_, err = iamSetPolicyCall.Do()
+	if err != nil {
+		return nil, "", err
+	}
+
+	keyRequest := &iam.CreateServiceAccountKeyRequest{}
+
+	key, err := m.iamService.Projects.ServiceAccounts.Keys.Create("projects/-/serviceAccounts/"+account.UniqueId, keyRequest).Do()
+	if err != nil {
+		return nil, "", err
+	}
+
+	saJson, err := base64.StdEncoding.DecodeString(key.PrivateKeyData)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return saJson, account.Email, err
+}
+
+func (m *Metabase) restore(ctx context.Context, datasetID uuid.UUID, mbMetadata *models.MetabaseMetadata) error {
+	ds, err := m.repo.GetBigqueryDatasource(ctx, datasetID)
+	if err != nil {
+		return err
+	}
+
+	err = m.accessMgr.Grant(ctx, ds.ProjectID, ds.Dataset, ds.Table, "serviceAccount:"+mbMetadata.SAEmail)
+	if err != nil {
+		m.log.Error("Unable to revoke access")
+		return err
+	}
+
+	if err := m.repo.RestoreMetabaseMetadata(ctx, datasetID); err != nil {
+		m.log.Error("Unable to soft create metabase metadata")
+		return err
+	}
+
+	return nil
+}
+
+func (m *Metabase) waitForDatabase(ctx context.Context, dbID int, tableName string) {
+	for i := 0; i < 50; i++ {
```

Comment:
5 seconds sounds a bit short in extreme cases... it involves remote request to remote database

Reply:
Increased it to ten seconds